mod/consensus/pkg/cometbft/service/abci.go

```diff
@@ -90,75 +89,76 @@ func (s *Service[LoggerT]) InitChain(
 		// InitChain represents the state of the application BEFORE the first
 		// block, i.e. the genesis block. This means that when processing the
 		// app's InitChain handler, the block height is zero by default.
-		// However, after Commit is called
-		// the height needs to reflect the true block height.
-		initHeader.Height = req.InitialHeight
+		// However, after genesis is handled the height needs to reflect
+		// the true block height.
+		initHeader := cmtproto.Header{
+			ChainID: req.ChainId,
+			Time:    req.Time,
+			Height:  req.InitialHeight,
+		}
 		s.finalizeBlockState.SetContext(
 			s.finalizeBlockState.Context().WithBlockHeader(initHeader),
 		)
 	}()
 
-	if s.finalizeBlockState == nil {
-		return nil, errors.New("finalizeBlockState is nil")
-	}
```

Comment:
redundant check since s.setState(execModeFinalize) is called right before. We don't do these checks after setState is called for proposals.